packages/flame_riverpod/lib/src/has_component_ref_mixin.dart

```diff
@@ -0,0 +1,44 @@
+import 'package:flame/components.dart';
+import 'package:flame_riverpod/src/component_ref.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+/// The [HasComponentRef] mixin is a bridge between a FlameGame and a
+/// ConsumerStatefulWidget. A single WidgetRef is shared between components
+/// within a FlameGame, accessed through the [ComponentRef] type. Subscriptions
+/// to providers are managed in accordance with the Component lifecycle.
+mixin HasComponentRef on Component {
+  ComponentRef get ref => _reference;
+  static late ComponentRef _reference;
+  static set widgetRef(WidgetRef value) => _reference = ComponentRef(value);
+
+  final List<ProviderSubscription> _subscriptions = [];
+
+  @override
+  Future<void> onLoad() async {
+    await super.onLoad();
+  }
+
+  void listen<T>(
+    ProviderListenable<T> provider,
+    void Function(T?, T) onChange, {
+    void Function(Object error, StackTrace stackTrace)? onError,
+  }) {
+    _subscriptions.add(
+      ref.listenManual<T>(
+        provider,
+        (p0, p1) {
+          onChange(p0, p1);
+        },
+        onError: onError,
+      ),
+    );
+  }
+
+  @override
+  void onRemove() {
+    for (final listener in _subscriptions) {
+      listener.close();
+    }
+    super.onRemove();
+  }
```

Comment:
What happens if the component is later re-added into the component tree?

Reply:
My understanding is that in that event, onMount is called once again.

Reply:
Yes, indeed. The question was more about what would happen to riverpod listener subscriptions -- will they get re-subscribed automatically? If not, then why are they being kept inside the `listeners` list?